subsys/net/lib/wifi_ready/wifi_ready.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <net/wifi_ready.h>
+#include <supp_events.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/net_event.h>
+
+LOG_MODULE_REGISTER(wifi_ready, CONFIG_WIFI_READY_EVENT_HANDLING_LOG_LEVEL);
+
+static wifi_ready_callback_t wifi_callbacks[CONFIG_MAX_CALLBACKS];
+static int callback_count;
+static K_MUTEX_DEFINE(wifi_mutex);
+
+#define WPA_SUPP_EVENTS (NET_EVENT_WPA_SUPP_READY) | \
+			(NET_EVENT_WPA_SUPP_NOT_READY)
+
+static struct net_mgmt_event_callback net_wpa_supp_cb;
+
+static void handle_wpa_supp_event(struct net_mgmt_event_callback *cb,
+				  struct net_if *iface, bool ready)
+{
+	if (ready) {
+		LOG_INF("Supplicant is ready for iface %p", iface);
+	} else {
+		LOG_INF("Supplicant is not ready for iface %p", iface);
+	}
+
+	k_mutex_lock(&wifi_mutex, K_FOREVER);
+	for (int i = 0; i < callback_count; i++) {
+		if (wifi_callbacks[i].iface == iface &&
+		    wifi_callbacks[i].wifi_ready_cb) {
+			wifi_callbacks[i].wifi_ready_cb(ready);
+		}
+	}
+	k_mutex_unlock(&wifi_mutex);
+}
+
+static void wpa_supp_event_handler(struct net_mgmt_event_callback *cb,
+	uint32_t mgmt_event, struct net_if *iface)
+{
+	LOG_DBG("Event received: %d", mgmt_event);
+	switch (mgmt_event) {
+	case NET_EVENT_WPA_SUPP_READY:
+		handle_wpa_supp_event(cb, iface, true);
+		break;
+	case NET_EVENT_WPA_SUPP_NOT_READY:
+		handle_wpa_supp_event(cb, iface, false);
+		break;
+	default:
+		LOG_DBG("Unhandled event (%d)", mgmt_event);
+		break;
+	}
+}
+
+void register_wifi_ready_callback(wifi_ready_callback_t cb, struct net_if *iface)
+{
+	k_mutex_lock(&wifi_mutex, K_FOREVER);
+
+	if (callback_count >= CONFIG_MAX_CALLBACKS) {
+		LOG_ERR("Failed to register callback, maximum count reached");
```

Comment:
```suggestion
		LOG_ERR("Reject callback registration, maximum count reached");
```

This function should return status and in this case it should be `ENOMEM` or something similar.